cardano-db-sync/src/Cardano/DbSync/AppT.hs

```diff
@@ -0,0 +1,294 @@
+  , soptCache :: !Bool
+  , soptSkipFix :: !Bool
+  , soptOnlyFix :: !Bool
+  , soptPruneConsumeMigration :: !DB.PruneConsumeMigration
+  , soptInsertOptions :: !InsertOptions
+  , snapshotEveryFollowing :: !Word64
+  , snapshotEveryLagging :: !Word64
+  }
+  deriving (Show)
+
+data InsertOptions = InsertOptions
+  { ioInOut :: !Bool
+  , ioUseLedger :: !Bool
+  , ioShelley :: !Bool
+  , ioRewards :: !Bool
+  , ioMultiAssets :: !Bool
+  , ioMetadata :: !Bool
+  , ioKeepMetadataNames :: Strict.Maybe [Word64]
+  , ioPlutusExtra :: !Bool
+  , ioOffChainPoolData :: !Bool
+  , ioGov :: !Bool
+  }
+  deriving (Show)
+
+-- A representation of if we are using a ledger or not given CLI options
+data LedgerEnv where
+  HasLedger :: HasLedgerEnv -> LedgerEnv
+  NoLedger :: NoLedgerEnv -> LedgerEnv
+
+data HasLedgerEnv = HasLedgerEnv
+  { leTrace :: Trace IO Text
+  , leUseLedger :: !Bool
+  , leHasRewards :: !Bool
+  , leProtocolInfo :: !(Consensus.ProtocolInfo CardanoBlock)
+  , leDir :: !LedgerStateDir
+  , leNetwork :: !Ledger.Network
+  , leSystemStart :: !SystemStart
+  , leAbortOnPanic :: !Bool
+  , leSnapshotEveryFollowing :: !Word64
+  , leSnapshotEveryLagging :: !Word64
+  , leInterpreter :: !(StrictTVar IO (Strict.Maybe CardanoInterpreter))
+  , leStateVar :: !(StrictTVar IO (Strict.Maybe LedgerDB))
+  , leStateWriteQueue :: !(TBQueue (FilePath, CardanoLedgerState))
+  }
+
+data NoLedgerEnv = NoLedgerEnv
+  { nleTracer :: Trace IO Text
+  , nleSystemStart :: !SystemStart
+  , nleQueryVar :: StateQueryTMVar CardanoBlock CardanoInterpreter
+  , nleHistoryInterpreterVar :: StrictTVar IO (Strict.Maybe CardanoInterpreter)
+  , nleNetwork :: !Ledger.Network
+  , nleProtocolInfo :: !(Consensus.ProtocolInfo CardanoBlock)
+  }
+
+newtype StateQueryTMVar blk result = StateQueryTMVar
+  { unStateQueryTMVar ::
+      StrictTMVar IO (Query blk result, StrictTMVar IO (Either AcquireFailure result))
+  }
+
+type RunMigration = DB.MigrationToRun -> IO ()
+
+data FixesRan = NoneFixRan | DataFixRan | AllFixRan
+
+data ConsistentLevel = Consistent | DBAheadOfLedger | Unchecked
+  deriving (Show, Eq)
+
+data EpochState = EpochState
+  { esInitialized :: !Bool
+  , esEpochNo :: !(Strict.Maybe EpochNo)
+  }
+
+--------------------------------------------------------------------------------
+-- Application Functions
+--------------------------------------------------------------------------------
+
+-- | Run an `App` action.
+runApp :: SyncEnv -> App a -> IO (Either SyncNodeError a)
+runApp env app = runExceptT (runReaderT (runAppT app) env)
+
+-- | Run an `App` action in `IO`.
+runAppInIO :: SyncEnv -> App a -> IO a
+runAppInIO syncEnv appM = do
+  result <- runApp syncEnv appM
+  case result of
+    Left err -> error $ "Error occurred: " ++ show err
+    Right val -> pure val
+
+-- | Lift an `App` action to fit the required type for `runDbIohkNoLogging`
+runAppWithNoLogging ::
+  MonadUnliftIO m =>
+  SyncEnv ->
+  App a ->
+  ReaderT SqlBackend (NoLoggingT m) a
```

Comment:
Another option here is to use the `SqlPersistT` [alias](https://hackage.haskell.org/package/persistent-2.14.6.1/docs/Database-Persist-Sql.html#t:SqlPersistT)

Reply:
Good catch, didn't even look for it

Reply:
could we do this as part of another PR just that all of the db queries are like this:

```
insertAdaPots :: (MonadBaseControl IO m, MonadIO m) => AdaPots -> ReaderT SqlBackend m AdaPotsId
insertAdaPots = insertUnchecked "AdaPots"

insertBlock :: (MonadBaseControl IO m, MonadIO m) => Block -> ReaderT SqlBackend m BlockId
insertBlock = insertUnchecked "Block"
```

Reply:
Yes, of course! Let's keep this PR as small as is reasonable